new_testing.Rmd

```diff
@@ -0,0 +1,201 @@
+---
+output: github_document
+---
+
+## New testing summary
+
+**Note: This document should be moved into the wiki once this PR is clear to be merged**
+
+
+This PR moves testing into the class methods, allowing tests to be ran interactively from within the country class object, and from within the existing testing infrastructure.
+
+`DataClass` is the parent class of all country classes, such as `Italy`, `UK`, `USA`, etc. The generic testing function is the method `test` within `DataClass`.
+
+Interactively, tests can be ran by doing the following:
+
+
+```{r}
+ukdata <- covidregionaldata::UK$new(level = "1", verbose = FALSE)
+# could use anything but I used the acutal test one here for simplicity
+ukdata$test(snapshot_path = "tests/testthat/custom_data/UK_level_1.rds")
+```
+
+
+Here, I break down the different components of this function to walk through how the tests are run.
+
+```{r, eval = FALSE, echo = TRUE}
+test <- function(download = FALSE, snapshot_path = "",
+                 test_all = FALSE, ...) {
+  if (!grepl(".rds$", snapshot_path)) {
+    stop("snapshot_path must be to an rds file")
+  }
+  self_copy <- self$clone()
+  test_download(
+    cntry_obj = self_copy,
+    download = download,
+    snapshot_path = snapshot_path
+  )
+  test_cleaning(
+    cntry_obj = self_copy
+  )
+  test_processing(
+    cntry_obj = self_copy,
+    test_all = test_all
+  )
+  test_return(
+    cntry_obj = self_copy
+  )
+
+  if ("specific_tests" %in% names(self_copy)) {
+    specific <- paste0(
+      "self$specific_tests(self_copy=self_copy, download=download, ...)"
+    )
+    eval(parse(text = specific))
+  }
+}
+```
+
+For a given country, such as the UK, you would make this by calling something like `ukdata <- UK$new(level = "1")` and can run the steps one by one by calling the respective methods: `ukdata$download(); ukdata$clean(); ukdata$process()`. To run the tests you would call `ukdata$test(shapshot_path = "path/2/data.rds")`. Rather than running tests on your active class, the code first makes a clone of your class which is then used for tests: `self_copy <- self$clone()` The snapshot path is the path to a rds file where you either have stored some raw downloaded data to test, or where you want a downloaded snapshot of the data to end up. This is handled by the function `test_download()`. The cloned class is passed to this function, along with the download parameter, which dictates whether to overwrite the snapshot file provided.
+
+```{r, eval = FALSE, echo = TRUE}
+test_download <- function(cntry_obj, download, snapshot_path) {
```

Comment:
I'd suggest loading this code in from the package vs copying it across to make this easier to update.